fenzo-core/src/main/java/com/netflix/fenzo/ScaleDownConstraintExecutor.java

```diff
@@ -1,38 +1,92 @@
 
 /**
- * Executes multiple {@link ScaleDownConstraintEvaluator} evaluators, and combines the results using {@link ScaleDownOrderResolver}.
+ * {@link ScaleDownConstraintExecutor} uses {@link ScaleDownOrderEvaluator} and multiple
+ * {@link ScaleDownConstraintEvaluator}s to determine termination order of candidate VM list. The evaluation process
+ * includes the following steps:
+ * <ul>
+ *     <li>{@link ScaleDownOrderEvaluator} splits candidate VMs into ordered list of equivalence groups</li>
+ *     <li>For each equivalence group multiple {@link ScaleDownOrderEvaluator}s are executed to produce a score for each VM</li>
+ *     <li>VMs are ordered in each equivalence group with the highest score first</li>
+ *     <li>The ordered equivalence groups are merged into single list, preserving the partitioning order computed by
+ *         {@link ScaleDownConstraintExecutor}. This list is returned as a result.
+ *     </li>
+ * </ul>
  */
 class ScaleDownConstraintExecutor {
 
     private static final Logger logger = LoggerFactory.getLogger(ScaleDownConstraintExecutor.class);
 
-    private final List<ScaleDownConstraintEvaluator> evaluators;
+    private static final double NOT_REMOVABLE_MARKER = -1;
+
+    private final ScaleDownOrderEvaluator orderEvaluator;
+    private final List<ScaleDownConstraintEvaluator> scoringEvaluators;
 
-    ScaleDownConstraintExecutor(List<ScaleDownConstraintEvaluator> evaluators) {
-        this.evaluators = evaluators;
+    ScaleDownConstraintExecutor(ScaleDownOrderEvaluator orderEvaluator, List<ScaleDownConstraintEvaluator> scoringEvaluators) {
+        this.orderEvaluator = orderEvaluator;
+        this.scoringEvaluators = scoringEvaluators;
     }
 
-    ScaleDownOrderResolver evaluate(Collection<VirtualMachineLease> candidates) {
-        List<ScaleDownConstraintEvaluator.Result> results = new ArrayList<>();
-        evaluators.forEach(evaluator -> {
-            try {
-                results.add(evaluator.evaluate(candidates));
-            } catch (Exception e) {
-                logger.info("Ignoring scale down constraint evaluator {} due to evaluation error", evaluator.getName(), e);
+    List<VirtualMachineLease> evaluate(Collection<VirtualMachineLease> candidates) {
+        List<Set<VirtualMachineLease>> fixedOrder = orderEvaluator.evaluate(candidates);
+
+        List<VirtualMachineLease> scaleDownOrder = scoringEvaluators.isEmpty()
+                ? fixedOrder.stream().flatMap(Set::stream).collect(Collectors.toList())
+                : fixedOrder.stream().flatMap(this::groupEvaluator).collect(Collectors.toList());
+
+        if (logger.isDebugEnabled()) {
+            List<String> hosts = scaleDownOrder.stream().map(VirtualMachineLease::hostname).collect(Collectors.toList());
+            logger.debug("Evaluated scale down order: {}", hosts);
+        }
+
+        return scaleDownOrder;
+    }
+
+    private Stream<VirtualMachineLease> groupEvaluator(Set<VirtualMachineLease> groupCandidates) {
```

Comment:
Would be good to have a test that exercises multiple evaluators with weighted scores.